lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp

```diff
@@ -1150,6 +1191,180 @@ void SESERegionBuilder::ensureSingleExitFromLoops() {
+  //     i2 = i1 + 1
+  //     cond ??, break, body
+  //
+  //   break:
+  //     br exit(i2)
+  //
+  //   body:
+  //     i3 = i2 + 1
+  //     cond ??, header(i3), exit(i3)
+  //
+  //   exit(i4):
+  //     return i4
+  //
+  // Note that the dataflow between iterations of the loop is captured by
+  // argument `i1` of the header. If we need to exit from the header of the
+  // loop, we will need to "freeze" the state at the current exit points and
+  // propagate it to the new header. The canonicalization pass does that by
+  // adding an argument to the header and passing it from the exit points. See
+  // argument `i5` of the header. We also have a `stayInLoop` argument to
+  // determine when we should exit the loop.
+  //
+  // The propagation of the frozen state is done as follows:
+  //   - The original branches to `exit` are replaced with branches to
+  //     `newLatch`, where `stayInLoop` is set to false.
+  //   - For the original argument `i1` at the header, we simply pass back that
+  //     value. This is OK because we exit the loop before accessing it.
+  //   - For the additional argument `i5` at the header, we pass the current
+  //     state at the exit.
+  //
+  // The canonicalized CFG is shown below:
+  //
+  //   preheader:
+  //     i0 = 0
+  //     br newHeader(i0, undef, true)
+  //
+  //   newHeader(i1, i4, stayInLoop):
+  //     cond stayInLoop, header, exit(i4)
+  //
+  //   header:
+  //     i2 = i1 + 1
+  //     cond ??, break, body
+  //
+  //   break:
+  //     br newLatch(i1, i2, false)
+  //
+  //   body:
+  //     i3 = i2 + 1
+  //     cond ??, newLatch(i3, i4, true), newLatch(i1, i3, false)
+  //
+  //   newLatch(i5, i6, stayInLoop1):
+  //     br newHeader(i5, i6, stayInLoop1)
+  //
+  //   exit(i7):
+  //     return i7
+  //
+  // We have an undef for `i5` from the preheader because we do not know the
+  // value before the loop gets executed. However, note that on all paths from
+  // `header` to `newLatch`, the second argument is always defined. Therefore,
+  // we can eliminate the `undef` by simply unrolling the loop body from
+  // `header` to `newLatch` once. We do not need to  clone `newHeader`.
+  //
+  // Given that we do not clone `newHeader`, references to the arguments `i1`
+  // and `i5` of `newHeader` will not be cloned as well. For example, in the
+  // unrolled body of the loop, `break` will be cloned as follows (prime refers
+  // to the clones):
+  //    break':
+  //      br newLatch'(i1, i2', false)
+  //
+  // Note that `i1` still refers to the original argument of `newHeader`.
+  // `newHeader` occurs after the unrolled loop body (in CFG) and hence, won't
+  // dominate `break'` in the unrolled loop body. We fix this by
+  // picking a value that is equivalent to the corresponding argument and
+  // dominates `break'`. (Note that it is enough to search for equivalent values
+  // among the immediate predecessors of the `newLatch'` block.)
+  //
+  // (1) Unroll the loop body once.
```

Comment:
this comment is confusing:
1. the comment block above so far talks about an example. Here we start explaining what the code does. But it's not clear to reader if this is still talking about the above example.

2. The comment coincides with the function name `unrollLoopBodyOnce()` -- so then why should there be subsequent steps (2), etc?